### ISSUE TITLE
Problem: deadlock between tendermint and abci server

### ISSUE THREAD

**Issue by sbellem** [bigchaindb/bigchaindb#2090]
### Problem
When tendermint is first started it tries to connect to the ABCI server. Until it establishes a connection with the ABCI server, calls to it such as requesting status information may fail.

`run_recover` tries to make a request to tendermint before the ABCI server is started and this causes errors, e.g.

```bash
$ docker-compose -f docker-compose.tendermint.yml up --no-deps bdb                  
Creating bigchaindb_bdb_1 ... done                              
Attaching to bigchaindb_bdb_1   
bdb_1          | INFO:bigchaindb.config_utils:Configuration loaded from `/root/.bigchaindb`                                      
bdb_1          | [2018-02-22 23:22:13] [INFO] (bigchaindb.commands.bigchaindb) BigchainDB Version 1.4.0.dev (MainProcess - pid: 1)
bdb_1          | [2018-02-22 23:22:13] [DEBUG] (bigchaindb.config_utils) System already configured, skipping autoconfiguration (MainProcess - pid: 1)
bdb_1          | [2018-02-22 23:22:13] [DEBUG] (bigchaindb.backend.connection) Connection: <class 'bigchaindb.backend.localmongodb.connection.LocalMongoDBConnection'> (MainProcess - pid: 1)
bdb_1          | [2018-02-22 23:22:13] [DEBUG] (urllib3.connectionpool) Starting new HTTP connection (1): tendermint (MainProcess - pid: 1)
bdb_1          | Traceback (most recent call last):             
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 141, in _new_conn                   
bdb_1          |     (self.host, self.port), self.timeout, **extra_kw)                                                           
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/util/connection.py", line 83, in create_connection       
bdb_1          |     raise err  
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/util/connection.py", line 73, in create_connection       
bdb_1          |     sock.connect(sa)                           
bdb_1          | ConnectionRefusedError: [Errno 111] Connection refused                                                          
bdb_1          |                
bdb_1          | During handling of the above exception, another exception occurred:                                             
bdb_1          |                
bdb_1          | Traceback (most recent call last):             
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen                 
bdb_1          |     chunked=chunked)                           
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request           
bdb_1          |     conn.request(method, url, **httplib_request_kw)                                                             
bdb_1          |   File "/usr/local/lib/python3.6/http/client.py", line 1239, in request                                         
bdb_1          |     self._send_request(method, url, body, headers, encode_chunked)                                              
bdb_1          |   File "/usr/local/lib/python3.6/http/client.py", line 1285, in _send_request                                   
bdb_1          |     self.endheaders(body, encode_chunked=encode_chunked)                                                        
bdb_1          |   File "/usr/local/lib/python3.6/http/client.py", line 1234, in endheaders                                      
bdb_1          |     self._send_output(message_body, encode_chunked=encode_chunked)                                              
bdb_1          |   File "/usr/local/lib/python3.6/http/client.py", line 1026, in _send_output                                    
bdb_1          |     self.send(msg)                             
bdb_1          |   File "/usr/local/lib/python3.6/http/client.py", line 964, in send                                             
bdb_1          |     self.connect()                             
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 166, in connect                     
bdb_1          |     conn = self._new_conn()                    
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connection.py", line 150, in _new_conn                   
bdb_1          |     self, "Failed to establish a new connection: %s" % e)                                                       
bdb_1          | urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f56d0935e10>: Failed to establish a new connection: [Errno 111] Connection refused
bdb_1          |                
bdb_1          | During handling of the above exception, another exception occurred:                                             
bdb_1          |                
bdb_1          | Traceback (most recent call last):             
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 440, in send                         
bdb_1          |     timeout=timeout                            
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen                 
bdb_1          |     _stacktrace=sys.exc_info()[2])             
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 388, in increment                   
bdb_1          |     raise MaxRetryError(_pool, url, error or ResponseError(cause))                                              
bdb_1          | urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='tendermint', port=46657): Max retries exceeded with url: /status (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f56d0935e10>: Failed to establish a new connection: [Errno 111] Connection refused',))
bdb_1          |                
bdb_1          | During handling of the above exception, another exception occurred:                                             
bdb_1          |                
bdb_1          | Traceback (most recent call last):             
bdb_1          |   File "/usr/local/bin/bigchaindb", line 11, in <module>                                                        
bdb_1          |     load_entry_point('BigchainDB', 'console_scripts', 'bigchaindb')()                                           
bdb_1          |   File "/usr/src/app/bigchaindb/commands/bigchaindb.py", line 377, in main                                      
bdb_1          |     utils.start(create_parser(), sys.argv[1:], globals())                                                       
bdb_1          |   File "/usr/src/app/bigchaindb/commands/utils.py", line 206, in start                                          
bdb_1          |     return func(args)                          
bdb_1          |   File "/usr/src/app/bigchaindb/commands/utils.py", line 51, in configure                                       
bdb_1          |     command(args)                              
bdb_1          |   File "/usr/src/app/bigchaindb/commands/utils.py", line 78, in start_logging                                   
bdb_1          |     command(args)                              
bdb_1          |   File "/usr/src/app/bigchaindb/commands/bigchaindb.py", line 196, in run_start                                 
bdb_1          |     run_recover(BigchainDB())                  
bdb_1          |   File "/usr/src/app/bigchaindb/commands/bigchaindb.py", line 162, in run_recover                               
bdb_1          |     tendermint_height = b.get_latest_block_height_from_tendermint()                                             
bdb_1          |   File "/usr/src/app/bigchaindb/tendermint/lib.py", line 54, in get_latest_block_height_from_tendermint         
bdb_1          |     r = requests.get(ENDPOINT + 'status')      
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 72, in get                                
bdb_1          |     return request('get', url, params=params, **kwargs)                                                         
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 58, in request                            
bdb_1          |     return session.request(method=method, url=url, **kwargs)                                                    
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request                      
bdb_1          |     resp = self.send(prep, **send_kwargs)      
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send                         
bdb_1          |     r = adapter.send(request, **kwargs)        
bdb_1          |   File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 508, in send                         
bdb_1          |     raise ConnectionError(e, request=request)  
bdb_1          | requests.exceptions.ConnectionError: HTTPConnectionPool(host='tendermint', port=46657): Max retries exceeded with url: /status (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f56d0935e10>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

meanwhile the tendermint server gives:

```bash
$ docker-compose -f docker-compose.tendermint.yml logs -f tendermint                
Attaching to bigchaindb_tendermint_1                            
tendermint_1   | E[02-22|23:21:55.029] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:08.036] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:16.046] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:19.048] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:22.050] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:25.052] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:28.054] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:31.057] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:34.059] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:37.061] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
tendermint_1   | E[02-22|23:22:40.064] abci.socketClient failed to connect to tcp://bdb:46658.  Retrying... module=abci-client connection=query
```
So both services are stalled as each one appears to depend on each other.

### Possible solutions
One possible solution would be to run the recovery step after the ABCI server is started as done in https://github.com/sbellem/bigchaindb/tree/run_recover-troubleshooting. See travis related job: https://travis-ci.org/sbellem/bigchaindb/jobs/345030105

**Comment by kansi** [comment]
I looked into this. The reason its crashing is because during `run_recover` a HTTP call is made to `http://tendermint:46657/status`  which fails. Following is a description of why it fails:

- Tendermint caches the `status` information when it starts up by calling `info` on the ABCI server.
- Once the above is successfully completed tendermint updates this `status` information during `begin_block` ... `commit` cycle.

**Scenairo 1** (Tendermint and BigchainDB both crash)
In this case when BigchainDB starts back up and executes `run_recover` the HTTP call to get `status` will crash because the `status` cache is lost once Tendermint shuts down. So, Tendermint needs to re-populate it by calling `info` on from ABCI server which hasn't yet started.

**Scenairo 2** (Tendermint is running BigchainDB crashes)
In this case when BigchainDB starts back up and executes `run_recover` the HTTP call to get `status` info will succeed because Tendermint will return the`status` from its cache even if the ABCI server is not running.


**Possible solution**
I think its better to just delete the latest block along with zombie transactions <sup>TM</sup> because if the recover approach depends on calling the tendermint HTTP api then **scenario 1** will not be covered.

**Comment by vrde** [comment]
The issue has been fixed by #2091. Recovering from crash is another problem.
